src/uct/ib/mlx5/ib_mlx5.inl

```diff
@@ -329,91 +337,103 @@ uct_ib_mlx5_set_ctrl_qpn_ds(struct mlx5_wqe_ctrl_seg *ctrl, uint32_t qp_num,
-                        8,               /* data size */
+                        4,               /* data size */
                         16,              /* signature (set 0) */
-                        16, 16,          /* reserved (set 0) */
+                        9, 8,            /* dci stream channel ID in BE */
                         0,               /* signal/fence_mode */
                         16, 16, 16, 16}; /* immediate (set to 0)*/
-    uint32x4_t data = {(opcode << 16) | (opmod << 8) | (uint32_t)fm_ce_se,
-                       pi, ds, qp_num};
+    uint32x4_t data = {(ds << 24) | (opcode << 16) | (opmod << 8) | fm_ce_se,
+                       pi, dci_channel, qp_num};
+#elif !defined (__SSE4_2__)
+    uct_ib_mlx5_wqe_ctrl_seg_t *uct_ctrl = (uct_ib_mlx5_wqe_ctrl_seg_t*)ctrl;
 #endif
 
     ucs_assert(((unsigned long)ctrl % UCT_IB_MLX5_WQE_SEG_SIZE) == 0);
+
 #if defined(__SSE4_2__)
     *(__m128i *) ctrl = _mm_shuffle_epi8(
-                    _mm_set_epi32(qp_num, ds, pi,
-                                  (opcode << 16) | (opmod << 8) | fm_ce_se), /* OR of constants */
-                    _mm_set_epi8(0, 0, 0, 0, /* immediate */
-                                 0,          /* signal/fence_mode */
-                                 0, 0,       /* reserved */
-                                 0,          /* signature */
-                                 8,          /* data size */
-                                 12, 13, 14, /* QP num */
-                                 2,          /* opcode */
-                                 4, 5,       /* sw_pi in BE */
-                                 1           /* opmod */
+                    _mm_set_epi32(qp_num, dci_channel, pi,
+                                  (ds << 24) | (opcode << 16) | (opmod << 8) | fm_ce_se), /* OR of constants */
+                    _mm_set_epi8(0x80, 0x80, 0x80, 0x80, /* immediate (set 0) */
+                                 0,                      /* signal/fence_mode */
+                                 8, 9,                   /* reserved */
+                                 0x80,                   /* signature (set 0) */
+                                 3,                      /* data size */
+                                 12, 13, 14,             /* QP num */
+                                 2,                      /* opcode */
+                                 4, 5,                   /* sw_pi in BE */
+                                 1                       /* opmod */
                                  ));
 #elif defined(__ARM_NEON)
     *(uint8x16_t *)ctrl = vqtbl1q_u8((uint8x16_t)data, table);
 #else
-    ctrl->opmod_idx_opcode = (opcode << 24) | (htons(pi) << 8) | opmod;
+    uct_ctrl->opmod_idx_opcode      = (opcode << 24) | (htons(pi) << 8) | opmod;
     uct_ib_mlx5_set_ctrl_qpn_ds(ctrl, qp_num, ds);
-    ctrl->fm_ce_se         = fm_ce_se;
+    uct_ctrl->dci_stream_channel_id = htons(dci_channel);
```

Comment:
can dci channel be more than uint8?
if not, endianess swap could be simpler

Reply:
so shall we change the parameter type of "dci_channel" from uint16_t to uint8_t?

Reply:
> so shall we change the parameter type of "dci_channel" from uint16_t to uint8_t?

ctrl segment accepts 16 bit value for dci channel, so it expects big endian value. we may calculate it by storing BE value in EP, but it will require one more byte to store

Reply:
> can you check if the compiler really eliminated the htons ?

it seems GCC has no ideas how to optimise this (called tons via plt section). added macro to convert byte into network datatype

btw, we may create own implementation for hton*** calls via compiler intrinsics